# Call Center Cleanup

For this assignment, we will be working with call center data. You can start working on the assignment after the first lesson on Exploratory Data Analysis. Make sure to read the whole assignment before starting anything! As you code along in the Jupyter notebook, you are asked to make note of the results of your analysis. Do so by clicking on the results box and adding your notes beside each question.

## Business Issue and Understanding

You are working for a company that has two call centers: the North Call Center and the South Call Center. The company is looking to possibly hire five additional reps to enhance customer experience. Your task is to explore how efficient the current reps are in each branch to determine which branch would benefit from additional assistance.

### How the Call Center Works

Call center representatives are assigned queues. When calls are assigned to a queue, the call is assigned to the next person in line in the queue. After a call is assigned to a representative, the amount of time between assignment and the call starting is divided into busy minutes and not ready minutes. If the call is incoming and a customer is waiting on the phone for a rep, the time is split into three categories: busy minutes, not ready minutes, and incoming call wait time. Once the rep has the customer on the phone, there might be during call wait time, where the call is put on hold while the rep gets an answer for the customer.

### Notes about the Dataset

If you haven't worked in a call center before, these notes might help you throughout your analysis.

* The call purpose is tagged for each call.
* The time of the call is tagged in 1 hour blocks starting at 9:00 AM and ending at 5:00 PM.
* Calls are tagged as incoming or outgoing.
* Reps are assigned to queues. When the reps are working, they take calls in the order of their queue.
* A call that is dropped due to technical error or missed by the center because they have reached maximum capacity is a lost call.
* An abandoned call is when the customer hangs up because they have been waiting for too long.
* Busy Minutes: the amount of time after a call comes in or needs to go out where the assigned rep is not available because they are busy with other customers.
* Not Ready Minutes: the amount of time after a call comes in or needs to go out where the assigned rep is not available because they are not ready (for example, getting water).
* Incoming Wait Time - amount of time after assigned rep is available to take the call customer waits for representative to pick up a call. This is tracked in seconds.
* During Call Wait Time - amount of time during call that customer has to wait for representative

## Getting Started

You have two CSVs at your disposal, `NorthCallCenter.csv` and `SouthCallCenter.csv`. Import the appropriate libraries and create two dataframes, one called `north_df` and one called `south_df`.

In [1]:
# Import the appropriate libraries with aliases
import pandas as pd
import matplotlib 
import matplotlib.pyplot as plt
import numpy as np
# Create two new dataframes

north_df = pd.read_csv("NorthCallCenter.csv")
south_df = pd.read_csv("SouthCallCenter.csv")

In [2]:
north_df.head(5)

Unnamed: 0 Branch     Call Purpose Time Block Incoming or Outgoing Queue  \
0           0  North        Complaint    9:00 AM             Incoming     B   
1           1  North    Sales Support   11:00 AM             Incoming     A   
2           2  North    Sales Support   11:00 AM             Incoming     B   
3           3  North    Sales Support    4:00 PM             Incoming     B   
4           4  North  Product Support   12:00 PM             Incoming     B   

  Rep ID Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0  Brent   NO          0          1             9                  2   
1    Cam   NO          0          0            11                  1   
2   Todd   NO          0          0             9                  3   
3  Brent   NO          0          0            11                  2   
4  Brent   NO          0          0             8                  2   

   Incoming Wait Time  During Call Wait Time  Calls  
0                 1.0                      4      1  
1                 1.0                      3      1  
2                 1.0                      3      1  
3                 1.0                      4      1  
4                 1.0                      4      1

In [3]:
print(north_df.shape)

(245, 15)


In [4]:
print(south_df.shape)

(314, 15)


In [5]:
south_df.head(5)

Unnamed: 0 Branch   Call Purpose Time Block Incoming or Outgoing Queue  \
0           0  South  Sales Support   10:00 AM             Incoming     D   
1           1  South  Sales Support   10:00 AM             Incoming     C   
2           2  South  Sales Support   10:00 AM             Incoming     C   
3           3  South  Sales Support   10:00 AM             Incoming     C   
4           4  South  Sales Support   12:00 PM             Incoming     C   

  Rep ID Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0   Kate   NO          0          0             9                  1   
1   Eric   NO          0          0             8                  2   
2  Susan   NO          0          0            10                  2   
3  Alice   NO          0          0            12                  1   
4  Sandy   NO          0          0             8                  3   

   Incoming Wait Time  During Call Wait Time  Calls  
0                 1.0                      2      1  
1                 1.0                      4      1  
2                 1.0                      4      1  
3                 1.0                      3      1  
4                 1.0                      3      1

<span style="background-color:orangered; color:orangered;">- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span>

## Exploratory Data Analysis

Time to do some EDA! In the process of learning more about the two datasets, answer the following questions. In the code blocks below enter the code used to solve the question.  At the end of the section record your answers by double clicking where asked.

#### EDA Question 1A:  How many reps are in the North branch?  

In [6]:
#Code Used:
north_df.value_counts("Rep ID")

Rep ID
Brent     37
Lilly     30
Duke      29
Todd      29
Joe       26
Cam       24
Xander    24
Amanda    23
Andy      23
dtype: int64

#### EDA Question 1B:  How many reps are in the South branch?  

In [7]:
#Code Used:
south_df.value_counts("Rep ID")

Rep ID
Eric      35
Randy     33
Sandy     32
Helga     31
George    29
Kate      29
Karl      28
Josh      26
Susan     26
Sharon    25
Alice     20
dtype: int64

#### EDA Question 2A:  What is the average busy minutes, not ready minutes, incoming wait time, and during call wait time for the North branch? 

In [8]:
#Code Used:
print("North Avg Busy Minutes:", north_df["Busy Minutes"].mean().round(2))
print("North Avg Not Ready Minutes:", north_df["Not Ready Minutes"].mean().round(2))
print("North Avg Incoming Wait Time:", north_df["Incoming Wait Time"].mean().round(2))
print("North Avg During Call Wait Time:", north_df["During Call Wait Time"].mean().round(2))

North Avg Busy Minutes: 9.99
North Avg Not Ready Minutes: 1.91
North Avg Incoming Wait Time: 3.05
North Avg During Call Wait Time: 2.97


#### EDA Question 2B:  What is the average busy minutes, not ready minutes, incoming wait time, and during call wait time for the South branch? 

In [9]:
#Code Used:
print("South Avg Busy Minutes:", south_df["Busy Minutes"].mean().round(2))
print("South Avg Not Ready Minutes:", south_df["Not Ready Minutes"].mean().round(2))
print("South Avg Incoming Wait Time:", south_df["Incoming Wait Time"].mean().round(2))
print("South Avg During Call Wait Time:", south_df["During Call Wait Time"].mean().round(2))

South Avg Busy Minutes: 10.05
South Avg Not Ready Minutes: 1.91
South Avg Incoming Wait Time: 3.0
South Avg During Call Wait Time: 3.08


#### EDA Question 3:  What is the number of calls taken for each time block(Both North and South combined)?

In [10]:
# Code Used:
north_df["Time Block"].value_counts()

12:00 PM    45
10:00 AM    43
2:00 PM     35
3:00 PM     30
4:00 PM     28
11:00 AM    25
1:00 PM     17
5:00 PM     17
9:00 AM      5
Name: Time Block, dtype: int64

In [11]:
south_df["Time Block"].value_counts()

12:00 PM    75
10:00 AM    56
3:00 PM     43
11:00 AM    31
2:00 PM     30
5:00 PM     26
4:00 PM     25
1:00 PM     23
9:00 AM      5
Name: Time Block, dtype: int64

In [12]:
Total_calls_per_block = north_df["Time Block"].value_counts() + south_df["Time Block"].value_counts()
Total_calls_per_block

10:00 AM     99
11:00 AM     56
12:00 PM    120
1:00 PM      40
2:00 PM      65
3:00 PM      73
4:00 PM      53
5:00 PM      43
9:00 AM      10
Name: Time Block, dtype: int64

##### Record your answers for the EDA section by double clicking here:
- 1a) Reps in the North Branch = 9
- 1b) Reps in the South Branch = 11

- 2a) North Branch, (round to the nearest hundredth):
> - Busy Minutes = 9.99
> - Not Ready Minutes = 1.91
> - Incoming Wait Time = 3.05
> - During Call Wait Time = 2.97
- 2b) South Branch, (round to the nearest hundredth):
> - Busy Minutes = 10.05
> - Not Ready Minutes = 1.91
> - Incoming Wait Time = 3.0
> - During Call Wait Time = 3.08
- 3) Total Calls taken for all branches per time block:
> - 9:00AM   = 10
> - 10:00AM  = 99
> - 11:00AM  = 56
> - 12:00PM  = 120
> - 1:00PM   = 40
> - 2:00PM   = 65
> - 3:00PM   = 73
> - 4:00PM   = 53
> - 5:00PM   = 43



<span style="background-color:orangered; color:orangered;">- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span>

## Cleaning Your Data 
Now you need to clean up the datasets. When cleaning the datasets, you may find that there isn't dirty data to clean. That is okay! Some questions you need to answer about the data sets.
* Add additional code blocks as needed to show how you came to your conclusions. Add comments in your code blocks to help others understand your thinking. 

#### Cleaning Question 1:  Is there any missing data and if yes explain what you would do with the missing data and why?
##### Record your answer by double clicking here:
> Your Answer: 

NORTH:missing values in incoming wait time- 66.53%

SOUTH:missing values in incoming wait time- 59.87%

Even though both of these are a significant amout of missing information, it missing could potentially have meaning. Maybe there was a 0 min wait time, thus it was not recorded. I will leave this column in the data for now. I will also need this information later to analyze their call center efficiency KPI.

In [13]:
# Code Used
# Add Additional Cells as needed.
for col in north_df.columns:
    pct_missing = np.mean(north_df[col].isnull())
    print('{} - {}%'.format(col, round(pct_missing*100,2)))
    
    #missing values in incoming wait time- 66.53%

Unnamed: 0 - 0.0%
Branch - 0.0%
Call Purpose - 0.0%
Time Block - 0.0%
Incoming or Outgoing - 0.0%
Queue - 0.0%
Rep ID - 0.0%
Sale - 0.0%
Lost Call - 0.0%
Abandoned - 0.0%
Busy Minutes - 0.0%
Not Ready Minutes - 0.0%
Incoming Wait Time - 66.53%
During Call Wait Time - 0.0%
Calls - 0.0%


In [14]:
for col in south_df.columns:
    pct_missing = np.mean(south_df[col].isnull())
    print('{} - {}%'.format(col, round(pct_missing*100,2)))
    
    #missing values in incoming wait time- 59.87%

Unnamed: 0 - 0.0%
Branch - 0.0%
Call Purpose - 0.0%
Time Block - 0.0%
Incoming or Outgoing - 0.0%
Queue - 0.0%
Rep ID - 0.0%
Sale - 0.0%
Lost Call - 0.0%
Abandoned - 0.0%
Busy Minutes - 0.0%
Not Ready Minutes - 0.0%
Incoming Wait Time - 59.87%
During Call Wait Time - 0.0%
Calls - 0.0%


#### Cleaning Question 2:  In the North data set, there are two separate "YES" values for a sale. Why and how did you rectify the issue?
##### Record your answer by double clicking here:
> Your Answer: 
If I want to accuratly be able to compare the conversion rate for sales, then I need to have all "YES"'s calculated together. I used str.strip() to strip the blank space/ whitespace in the "Sale" column. I then recounted the value_counts to be sure it worked.

In [15]:
north_df["Sale"].value_counts()

NO      234
YES      10
YES       1
Name: Sale, dtype: int64

In [16]:
#Code Used
north_df['Sale'] = north_df['Sale'].str.strip()


In [17]:
north_df["Sale"].value_counts()

NO     234
YES     11
Name: Sale, dtype: int64

In [18]:
north_df.head(2)

Unnamed: 0 Branch   Call Purpose Time Block Incoming or Outgoing Queue  \
0           0  North      Complaint    9:00 AM             Incoming     B   
1           1  North  Sales Support   11:00 AM             Incoming     A   

  Rep ID Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0  Brent   NO          0          1             9                  2   
1    Cam   NO          0          0            11                  1   

   Incoming Wait Time  During Call Wait Time  Calls  
0                 1.0                      4      1  
1                 1.0                      3      1

#### Cleaning Question 3:  Are there duplicates in the two data sets? If there are, how did you handle them?
##### Record your answer by double clicking here:
> Your Answer: 
After taking out the "Unnamed: 0" because I knew it would be unique, I found the amount of dupplicates for each dataframe.
NORTH: 2 dupplicates removed
SOUTH: 3 dupplicates removed

In [19]:
#Code Used
# Add Additional Cells as needed.
north_df_dedupped = north_df.drop('Unnamed: 0', axis=1).drop_duplicates()

# there were duplicate rows
print("north_df shape:", north_df.shape)
print("north_df_dedupped shape:", north_df_dedupped.shape)
# df = df_dedupped

south_df_dedupped = south_df.drop('Unnamed: 0', axis=1).drop_duplicates()

# there were duplicate rows
print("south_df shape:", south_df.shape)
print("south_df_dedupped shape:", south_df_dedupped.shape)


north_df shape: (245, 15)
north_df_dedupped shape: (243, 14)
south_df shape: (314, 15)
south_df_dedupped shape: (311, 14)


#### Cleaning Question 4:  Is any of the data in the two data sets unnecessary? If yes, how did you handle it?
##### Record your answer by double clicking here:
> Your Answer: 
The unnamed column is not necessary becuase we already have an index shown. I removed that with my dedupped df above.
The column "Calls" seems unneccesary- all values are 1 becuase each call is stored separetely. I chose not to remove this column though so I could use it as a sort key and count with it later in my analysis.

In [20]:
#Code Used
# Add Additional Cells as needed.
# north_df["Calls"].value_counts()

In [21]:
# south_df["Calls"].value_counts()

In [22]:
# north_df = north_df_dedupped.drop('Calls', axis=1)
# south_df = south_df_dedupped.drop('Calls', axis=1)

#chose not to take out Calls so I could count using that column later

In [23]:
north_df = north_df_dedupped
south_df = south_df_dedupped

In [24]:
north_df.head()

Branch     Call Purpose Time Block Incoming or Outgoing Queue Rep ID Sale  \
0  North        Complaint    9:00 AM             Incoming     B  Brent   NO   
1  North    Sales Support   11:00 AM             Incoming     A    Cam   NO   
2  North    Sales Support   11:00 AM             Incoming     B   Todd   NO   
3  North    Sales Support    4:00 PM             Incoming     B  Brent   NO   
4  North  Product Support   12:00 PM             Incoming     B  Brent   NO   

   Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  Incoming Wait Time  \
0          0          1             9                  2                 1.0   
1          0          0            11                  1                 1.0   
2          0          0             9                  3                 1.0   
3          0          0            11                  2                 1.0   
4          0          0             8                  2                 1.0   

   During Call Wait Time  Calls  
0                      4      1  
1                      3      1  
2                      3      1  
3                      4      1  
4                      4      1

<span style="background-color:orangered; color:orangered;">- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span>

## Data Manipulation
Before you begin answering any questions, combine the two datasets together to create a third dataframe called df. You can use this third dataframe to compare the two call centers to company-wide trends.
* Add additional code blocks as needed to show how you came to your conclusions. Add comments in your code blocks to help others understand your thinking. Record your answer by double clicking the below cell.

In [25]:
# Create dataframe for the entire company named df
df = pd.concat([north_df, south_df], ignore_index = True)
df.shape

(554, 14)

In [26]:
df.head(2)
#added to show top of df- hoping it was north's data then will run tail below to be sure that south's data is there

Branch   Call Purpose Time Block Incoming or Outgoing Queue Rep ID Sale  \
0  North      Complaint    9:00 AM             Incoming     B  Brent   NO   
1  North  Sales Support   11:00 AM             Incoming     A    Cam   NO   

   Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  Incoming Wait Time  \
0          0          1             9                  2                 1.0   
1          0          0            11                  1                 1.0   

   During Call Wait Time  Calls  
0                      4      1  
1                      3      1

In [27]:
df.tail(2)

Branch     Call Purpose Time Block Incoming or Outgoing Queue  Rep ID  \
552  South  Product Support    5:00 PM             Outgoing     C    Eric   
553  South  Product Support    5:00 PM             Outgoing     D  George   

    Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
552   NO          0          0            10                  1   
553   NO          0          0            10                  2   

     Incoming Wait Time  During Call Wait Time  Calls  
552                 NaN                      2      1  
553                 NaN                      3      1

#### Manipulation Question 1:  Group by Rep ID and sum the resulting structure. Sort by calls to determine which rep in each branch has the highest number of calls.

##### Record your answer by double clicking here:
Rep with the hightest number of calls and their total calls:
- North Branch = Brent- 37 Calls
- South Branch = Eric - 35 Calls


In [28]:
# Code Used
# Add Additional Cells as needed.
df_grouped = df.groupby("Rep ID")
# df_grouped["Rep ID"].value_counts()
df_grouped.first().sort_values("Branch")

Branch     Call Purpose Time Block Incoming or Outgoing Queue Sale  \
Rep ID                                                                      
Joe     North  Product Support    2:00 PM             Incoming     A   NO   
Lilly   North  Product Support    2:00 PM             Incoming     B   NO   
Todd    North    Sales Support   11:00 AM             Incoming     B   NO   
Duke    North  Product Support    2:00 PM             Incoming     B   NO   
Xander  North  Product Support    3:00 PM             Incoming     B   NO   
Brent   North        Complaint    9:00 AM             Incoming     B   NO   
Andy    North        Complaint    2:00 PM             Incoming     A   NO   
Amanda  North  Product Support    4:00 PM             Incoming     A   NO   
Cam     North    Sales Support   11:00 AM             Incoming     A   NO   
Eric    South    Sales Support   10:00 AM             Incoming     C   NO   
George  South  Product Support    3:00 PM             Incoming     D   NO   
Helga   South  Product Support    3:00 PM             Incoming     D   NO   
Josh    South  Product Support    4:00 PM             Incoming     D   NO   
Karl    South    Sales Support    1:00 PM             Incoming     D   NO   
Kate    South    Sales Support   10:00 AM             Incoming     D   NO   
Randy   South  Product Support    2:00 PM             Incoming     D   NO   
Sandy   South    Sales Support   12:00 PM             Incoming     C   NO   
Sharon  South  Product Support   11:00 AM             Incoming     C   NO   
Susan   South    Sales Support   10:00 AM             Incoming     C   NO   
Alice   South    Sales Support   10:00 AM             Incoming     C   NO   

        Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
Rep ID                                                          
Joe             0          0             8                  1   
Lilly           0          0            11                  2   
Todd            0          0             9                  3   
Duke            0          0            12                  2   
Xander          0          0            12                  1   
Brent           0          1             9                  2   
Andy            0          1             9                  3   
Amanda          0          0            10                  3   
Cam             0          0            11                  1   
Eric            0          0             8                  2   
George          0          0             8                  3   
Helga           0          0             9                  1   
Josh            0          0            11                  3   
Karl            0          0             9                  1   
Kate            0          0             9                  1   
Randy           0          0            11                  2   
Sandy           0          0             8                  3   
Sharon          0          0            10                  1   
Susan           0          0            10                  2   
Alice           0          0            12                  1   

        Incoming Wait Time  During Call Wait Time  Calls  
Rep ID                                                    
Joe                    1.0                      3      1  
Lilly                  1.0                      3      1  
Todd                   1.0                      3      1  
Duke                   1.0                      2      1  
Xander                 2.0                      4      1  
Brent                  1.0                      4      1  
Andy                   2.0                      4      1  
Amanda                 1.0                      2      1  
Cam                    1.0                      3      1  
Eric                   1.0                      4      1  
George                 1.0                      4      1  
Helga                  1.0                      3      1  
Josh                   1.0                      3      

In [29]:
df_grouped.agg("sum")

C:\Users\longs\AppData\Local\Temp\ipykernel_2388\1413511487.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df_grouped.agg("sum")


Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
Rep ID                                                          
Alice           0          0           199                 46   
Amanda          0          1           218                 41   
Andy            0          3           228                 43   
Brent           0          1           362                 71   
Cam             1          0           233                 46   
Duke            2          0           292                 56   
Eric            0          0           361                 54   
George          0          0           298                 53   
Helga           0          2           292                 58   
Joe             0          0           246                 48   
Josh            3          0           260                 50   
Karl            1          0           287                 61   
Kate            0          0           283                 60   
Lilly           0          2           307                 59   
Randy           3          1           312                 63   
Sandy           1          0           317                 59   
Sharon          1          1           257                 38   
Susan           2          0           267                 53   
Todd            1          0           296                 58   
Xander          1          0           244                 44   

        Incoming Wait Time  During Call Wait Time  Calls  
Rep ID                                                    
Alice                 15.0                     58     20  
Amanda                12.0                     58     22  
Andy                  36.0                     65     23  
Brent                 17.0                    117     37  
Cam                   32.0                     66     24  
Duke                  39.0                     85     29  
Eric                  47.0                    112     35  
George                41.0                     87     29  
Helga                 36.0                     95     31  
Joe                   32.0                     78     25  
Josh                  31.0                     83     26  
Karl                  26.0                     85     28  
Kate                  34.0                     80     28  
Lilly                 37.0                     88     30  
Randy                 47.0                     96     32  
Sandy                 34.0                     96     31  
Sharon                29.0                     77     25  
Susan                 35.0                     88     26  
Todd                  21.0                     97     29  
Xander                20.0                     68     24

In [30]:
df_grouped.agg({"Calls" : "sum"}).sort_values("Calls", axis = 0, ascending = False, inplace = False)

Calls
Rep ID       
Brent      37
Eric       35
Randy      32
Sandy      31
Helga      31
Lilly      30
Todd       29
Duke       29
George     29
Karl       28
Kate       28
Josh       26
Susan      26
Joe        25
Sharon     25
Xander     24
Cam        24
Andy       23
Amanda     22
Alice      20

#### Manipulation Question 2:  The average call abandonment rate is a KPI when it comes to determining call center efficiency. As you may recall, abandoned calls are calls where the customer hangs up due to long call times. What is the average call abandonment rate for each branch and the whole company? Do any of these fall out of the optimal range of 2-5%?
##### Record your answer by double clicking here:
Average Call Abandonment Rates (round to the nearest hundredth):
- North Branch = 2.88%
- South Branch = 1.29%
- Company Wide = 1.99%
- Do any of these fall out of the optimal range of 2-5%?
None of them fall outside of the optimal range. North has the higher abandoned call average.

In [55]:
# Code Used  
df_abandonded_avg = df.agg({"Abandoned" : "sum"})/len(df)*100
df_abandonded.round(2)

Abandoned    1.99
dtype: float64

In [76]:
north_df_abandonded_avg = north_df.agg({"Abandoned" : "sum"})/len(north_df)*100
north_df_abandonded_avg.round(2)

Abandoned    2.88
dtype: float64

In [75]:
south_df_abandonded_avg = south_df.agg({"Abandoned" : "sum"})/len(south_df)*100
south_df_abandonded_avg.round(2)

Abandoned    1.29
dtype: float64

In [92]:
#this also works
# s = south_df.agg({"Abandoned" : "mean"})
# s.round(4)

# s = south_df["Abandoned"].mean()*100
# s.round(2)

1.29

#### Manipulation Question 3:  Service level is another KPI when it comes to measuring call center efficiency. Service level is the percentage of calls answered within a specific number of seconds. In the case of your employer, their ideal time frame is 2 seconds. What is the percentage of calls answered within 2 seconds for each branch and the entire company?
##### Record your answer by double clicking here:
Percentage of calls answered within 2 seconds, include # of calls:
- North Branch = 33 calls answered within 2 seconds--> 13.58%
- South Branch = 48 calls answered within 2 seconds--> 15.43%
- Company Wide = 81 calls answered within 2 seconds--> 14.62%

In [99]:
# Code Used
# # Add Additional Cells as needed.

north_wait2 =0
for value in north_df["Incoming Wait Time"]:
    if value <=2:
        north_wait2 +=1
    else:
        north_wait2 += 0

print("North_wait2 =", north_wait2)
print("North calls answered in under 2 seconds on average:", (north_wait2/len(north_df))*100)

North_wait2 = 33
North calls answered in under 2 seconds on average: 13.580246913580247


In [109]:
# north_wait2.mean().round(4)
# #IS THERE A WAY TO FIND THE MEAN AS A PERCENT AND ROUND IN THE SAME CHAINED SYNTAX???

0.1358

In [102]:
north_wait2 = north_df["Incoming Wait Time"]<=2
north_wait2.value_counts()

False    210
True      33
Name: Incoming Wait Time, dtype: int64

In [105]:
south_wait2 =0
for value in south_df["Incoming Wait Time"]:
    if value <=2:
        south_wait2 +=1
    else:
        south_wait2 += 0

print("south_wait2 =", south_wait2)
print("South calls answered in under 2 seconds on average:", (south_wait2/len(south_df))*100)

south_wait2 = 48
South calls answered in under 2 seconds on average: 15.434083601286176


In [106]:
df_wait2 =0
for value in df["Incoming Wait Time"]:
    if value <=2:
        df_wait2 +=1
    else:
        df_wait2 += 0

print("df_wait2 =", df_wait2)
print("Company calls answered in under 2 seconds on average:", (df_wait2/len(df))*100)

df_wait2 = 81
Company calls answered in under 2 seconds on average: 14.620938628158845


#### Manipulation Question 4: For each branch and the entire company, what is the average speed of answer?
##### Record your answer by double clicking here:
Average speed of answer (rounded to nearest hundredth):
- North Branch in seconds = 3.04 seconds
- South Branch in seconds = 3.0 seconds
- Company Wide in seconds = 3.01 seconds


In [80]:
# Code Used 
north_df["Incoming Wait Time"].describe()

count    81.000000
mean      3.037037
std       1.461544
min       1.000000
25%       2.000000
50%       3.000000
75%       4.000000
max       5.000000
Name: Incoming Wait Time, dtype: float64

In [82]:
north_df["Incoming Wait Time"].mean().round(2)

3.04

In [83]:
south_df["Incoming Wait Time"].mean().round(2)

3.0

In [84]:
df["Incoming Wait Time"].mean().round(2)

3.01

<span style="background-color:orangered; color:orangered;">- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span>

## Visualization

Create a visualization for each of the following questions. Some of the code to handle aggregating and storing data may be written for you. For each visualization, you choose the chart style that you feel suits the situation best. Make note of the chart style you chose and why. 

*NOTE Some questions you may decide to use more than one chart and or chart style.

#### Visualization 1:  What is the average abandonment rate per queue?
##### Record your answer by double clicking here:

> Chart style you chose and why: 

In [ ]:
# Create visualization 1 here
# Import any additional libraries needed with alias



# The dictionary abandonment_rates has the data you need.
abandonment_rates = {}
queues = ["A", "B", "C", "D"]
queue_dict = df.groupby("Queue").agg("sum")
for i in range(4):
    abandonment_rates[queues[i]] = queue_dict["Abandoned"][i] / queue_dict["Calls"][i] 

#Your Code Here:




#### Visualization 2: What is the service level percentage and average speed of answer for each rep in the North Branch?
##### Record your answer by double clicking here:

> Chart style you chose and why: 

In [ ]:
# Create visualization 2 here
# north_plt contains the data you need for the average speed of answer of each rep

north_plt = north.groupby("Rep ID")["Incoming Wait Time"].mean().to_frame().reset_index()

# Finding each Rep's Personal Service Level Percentage.  Basically, Calls within 2 secs / total calls

# Table 1: Total Incoming calls less than 2 seconds grouped by Rep
quick_calls = north[north["Incoming Wait Time"] <= 2.0]
quick_reps = quick_calls[["Rep ID", "Calls"]]
quick_stats = quick_reps.groupby(["Rep ID"]).sum()  # Final Table


# Table 2: Total Incoming Calls Only grouped by Rep
total_calls_in = north[north["Incoming or Outgoing"] == "Incoming"]
rep_calls = total_calls_in[["Rep ID", "Calls"]]     
total_stats = rep_calls.groupby(["Rep ID"]).sum() # Final Table  

#  Table 3: Service Level Percentage created via merge
service_level = pd.merge(quick_stats, total_stats, on="Rep ID")

# Create Percentage Column in Table 3
service_level["Service Level %"] = service_level["Calls_x"]/service_level["Calls_y"] * 100

#Your Code Here:


#### Visualization 3: For each type of call purpose, how many calls are outgoing vs. incoming?
##### Record your answer by double clicking here:

> Chart style you chose and why: 

In [ ]:
# Create visualization 3 here:
# The three dictionaries, complaints, sales_support, and product_support, have the information you need

purpose_group = df.groupby("Call Purpose")
call_purpose = ["Complaint", "Product Support", "Sales Support"]
purpose_counts = purpose_group["Incoming or Outgoing"].value_counts()
print(purpose_counts)

complaints = purpose_counts["Complaint"].to_dict()
sales_support = purpose_counts["Sales Support"].to_dict()
product_support = purpose_counts["Product Support"].to_dict()

#Your Code Here:


<span style="background-color:orangered; color:orangered;">- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span>

## Summarize Your Work

With what you know now about the two call centers and the entire company, answer the following questions. Note that while this is subjective, you should include relevant data to back up your opinion.


#### Summary Question 1:  Using KPIs such as average abandonment rate, service level and average speed of answer, in your opinion, which one of the two branches is operating more efficiently? Why?

##### Record your answer by double clicking here:
> Your Answer: 



#### Summary Question 2: Based on the number of reps in each branch and how quickly the reps are working, in your opinion, which branch would benefit from the extra help?

##### Record your answer by double clicking here:
> Your Answer: 


#### Summary Question 3: Now that you have explored the datasets, is there any data or information that you wish you had in this analysis?

##### Record your answer by double clicking here:
> Your Answer: 


<span style="background-color:orangered; color:orangered;">- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span>

## Bonus Mission
Create a visualization that answers this question: For each call purpose, how many calls (incoming and outgoing) take place in each time block?
##### Record your answer by double clicking here:

> Chart style you chose and why: 

In [ ]:
# Create your Bonus Mission visualization here!
call_times = df[["Time Block", "Call Purpose", "Incoming or Outgoing", "Calls"]]

# Use groupby to plot based on time blocks:

# Use groupby and get_group to select which call purpose to plot: